### DISCUSSION TITLE
Emoji Rules

### DISCUSSION THREAD

**Discussion by Ducky6944** [wezterm/wezterm#1037] [Q&A]
I am wondering if I can edit how emojis are displayed. Comparing the two screenshots below (wezterm and kitty) you can see that in wezterm the emojis appear below the horizontal line that the text sits on. They also seem larger and I'm not sure why but the two snake emojis are different. I tried looking through font related settings and fallback rules, but couldn't find any settings which might resolve this. 

<img width="499" alt="Screen Shot 2021-08-17 at 12 00 36 PM" src="https://user-images.githubusercontent.com/12100960/129760720-c371feba-c263-4a39-b6d2-e6d065d913b3.png">

<img width="479" alt="Screen Shot 2021-08-17 at 12 01 06 PM" src="https://user-images.githubusercontent.com/12100960/129760713-ac120724-e38b-42ce-8e0f-b7b81f787434.png">

**Comment by kar1** [comment]
The snake emoji is clearly from two different fonts. Can you post your font settings from Kitty and Wezterm and the output from `wezterm ls-fonts`. That would help diagnose the problem.

You can get the name of the font Wezterm is using by running the command:

```sh
wezterm ls-fonts --text '🐍'
```

**Reply by Ducky6944** [reply]
That makes total sense, but why? 
$`wezterm ls-fonts --text '🐍'`
```SHELL
\u{1f40d}    glyph=505  wezterm.font("Noto Color Emoji", {weight="Regular", stretch="Normal", italic=false})
```

Shouldn't it use the font specified in my config for emojis?

Wezterm Config
```LUA
local wezterm = require 'wezterm';
local home = wezterm.home_dir;
wezterm.add_to_config_reload_watch_list(home.."/.cache/wal/wezterm-wal.toml")
return {
  color_scheme_dirs = {home.."/.cache/wal"},
  hide_tab_bar_if_only_one_tab = true,
  color_scheme = "wezterm-wal",
  font = wezterm.font("Iosevka Nerd Font"),
  font_size = 14.0,
  freetype_interpreter_version = 35,
  send_composed_key_when_left_alt_is_pressed=false,
  send_composed_key_when_right_alt_is_pressed=true,
  use_ime = false,
  window_decorations = "RESIZE",
  window_padding = {
    left = 12,
    -- This will become the scrollbar width if you have enabled the scrollbar!
    right = 12,
    top = 12,
    bottom = 0,
  },
  leader = { key="a", mods="CTRL", timeout_milliseconds=1000 },
}

```

Kitty Config
```CONF
# vim:fileencoding=utf-8:ft=conf
# K I T T Y 

# C O L O R S 
include $HOME/.cache/wal/colors-kitty.conf
# Lighter Background
# background #333333

# F O N T S
font_size 16.0
font_family Iosevka Nerd Font
# italic_font      auto
# bold_font        auto
# bold_italic_font auto
disable_ligatures never

# T A B S
tab_bar_edge top
tab_bar_style powerline
tab_powerline_style angled
tab_bar_margin_height 1.0 1.0

# W I N D O W
hide_window_decorations titlebar-only
window_padding_width 5 5 5 5
inactive_text_alpha 0.6
mouse_hide_wait 3.0

# M A C O S
macos_option_as_alt both
macos_custom_beam_cursor yes

# S E S S I O N
allow_remote_control yes
cursor_shape beam
enable_audio_bell no

# K E Y B O A R D
# switch between next and previous splits
map cmd+]        next_window
map cmd+[        previous_window
# clear the terminal screen
map cmd+k combine : clear_terminal scrollback active : send_text normal,application \x0c
# jump to beginning and end of word
map alt+left send_text all \x1b\x62
map alt+right send_text all \x1b\x66
# jump to beginning and end of line
map cmd+left send_text all \x01
map cmd+right send_text all \x05

# S P L I T S
enabled_layouts fat:bias=50;full_size=1;mirrored=true, vertical, horizontal, grid, stack, splits:split_axis=horizontal, fat:bias=50;full_size=1;mirrored=false, tall:bias=50;full_size=1;mirrored=false

map F5 launch --location=hsplit
map F6 launch --location=vsplit
map F7 layout_action rotate

map shift+up move_window up
map shift+left move_window left
map shift+right move_window right
map shift+down move_window down

map shift+ctrl+left neighboring_window left
map shift+ctrl+right neighboring_window right
map shift+ctrl+up neighboring_window up
map shift+ctrl+down neighboring_window down

draw_minimal_borders no
inactive_border_color #cccccc
active_border_color #00ff00
window_border_width 2.5pt

```

**Reply by kar1** [reply]
To me it looks like Wezterm is using a Google font and Kitty is using a Apple font. Se here: https://emojipedia.org/snake/

Try specifying the apple font earlier in the fallback list. Wezterm will fallback to the included Google Noto Emoji font if the user doesn't request another font.

**Reply by Ducky6944** [reply]
Using  `wezterm.font_with_fallback({"Iosevka Nerd Font", "Apple Color Emoji"}),` resolves the different emojis, but they are still beneath the line that the characters are on. 
<img width="537" alt="Screen Shot 2021-08-17 at 1 46 03 PM" src="https://user-images.githubusercontent.com/12100960/129775384-0271144c-58a6-46bb-9cbf-23bd7d87b843.png">

**Reply by kar1** [reply]
Hmm... I can't reproduce it here, I'm on Debian. We'll have to wait for one of the Apple guys.
Thank you for the cooperation, we got half the way :-)

**Reply by Ducky6944** [reply]
No worries, a win is a win. I'm 0 for 2 on diagnosing what should be pretty easy issues on wezterm lol, so I appreciate the help.

**Reply by wez** [reply]
> Shouldn't it use the font specified in my config for emojis?

Noto Color Emoji is bundled with wezterm and is used as an implied fallback font, so it is using the fonts specified in your config (as revealed by `wezterm ls-fonts`) to lookup the snake.
 
If you try to display a glyph that isn't found in the explicit + implicit fallback fonts, wezterm will ask the system for its list of fallback fonts and then analyze those to figure out whether they contain the glyph.  That initial system fallback resolve is not cheap (we run it async in another thread and it can take visible fractions of a second to produce a match), but that is where wezterm would otherwise discover the Apple Color Emoji font.

If you prefer the apple font, then you can explicitly list it in your config to have it take precedence over the implied Noto font.

Re: line height: that's a tricky area.  You may want to try setting `use_cap_height_to_scale_fallback_fonts = true` in your config to see if things line up better.  Whether they do or not depends on how accurate the *cap-height* metric is for both your base font and the emoji font.   If you want to understand what influences that, take a look at https://tonsky.me/blog/font-size/

Another wrinkle with emoji fonts, aside from not having easily relatable font metrics, is that the glyphs are typically square and oversized relative to monospace terminals and need to be scaled to fit the box, and that scaling can mess with the line height.

So far no one has been sufficiently bothered by it to do something about it, but it may not be that much effort to reconcile for someone with the motivation and the time!

**Reply by Ducky6944** [reply]
Hey thanks for the response, I had never even realized that I guess that terminals we're even using separate emoji fonts. I realized shortly after I said that I was an idiot again lol. I tried the `use_cap_height_to_scale_fallback_fonts = true` to no avail. It would seem to me that they scale is appropriate, but where the character start at the very bottom of the line even lower than the descenders of the letters. Though this may be what your referring to when you say scaling messes with line height. I don't have the skill, but I definitely have the time. Let's see if I can acquire the skill and contribute something. Thanks for post, I'll do some research. 

<img width="178" alt="Screen Shot 2021-08-17 at 7 10 22 PM" src="https://user-images.githubusercontent.com/12100960/129812260-52675f08-3c18-405d-89dc-54f6ab26cb9c.png">

<img width="429" alt="Screen Shot 2021-08-17 at 7 06 30 PM" src="https://user-images.githubusercontent.com/12100960/129811956-ad923e09-f067-48c8-b69b-86c29a2aceb6.png">

**Reply by wez** [reply]
Don't worry, I didn't know anything useful about fonts until I started this project :-p
If you wanted to tool around with the code to get a feel, this is a good place to start:

https://github.com/wez/wezterm/blob/main/wezterm-gui/src/glyphcache.rs#L420-L431

That logic sets a scale for the glyph based on the ratio of metrics of the emoji font vs. the main font you specified.
If scale == 1.0 then we don't change the size, but if we do, this bit of code is triggered:

https://github.com/wez/wezterm/blob/main/wezterm-gui/src/glyphcache.rs#L505-L516

The data produced there is eventually used over here in the renderer:

https://github.com/wez/wezterm/blob/main/wezterm-gui/src/termwindow/render.rs#L1163-L1179

It's possible/probable that this particular line may need to be tweaked in a different way to accommodate
the scaling and get a more reasonable line height; right now it scales based solely on the cell size:

https://github.com/wez/wezterm/blob/main/wezterm-gui/src/termwindow/render.rs#L1176

but it may make sense to somehow account for differences in the descender metric that is also
available in those render metrics:

https://github.com/wez/wezterm/blob/main/wezterm-gui/src/utilsprites.rs#L13

For more information on what the font metrics terminology means, this page is pretty good:
https://www.freetype.org/freetype2/docs/tutorial/step2.html

**Reply by wez** [reply]
btw: one thing I found surprising when I first got exposed to this stuff was that emoji fonts are typically bitmap fonts that use fixed size 128x128 pixel images. Re: my commentary about the descender metric, I think the top/bottom position might need to be adjusted so that the descender value for the emoji font, with the scaling/font size adjustment accounted for, matches the descender value for the base font.  I think the math is probably trivial if you have the time to sit down and play with it and figure it out.

**Reply by Ducky6944** [reply]
Awesome, Thanks for the write up and references! I will see what I can do.
